units/log_output_tests.cc

```diff
@@ -0,0 +1,179 @@
+// limitations under the License.
+
+#include <array>
+#include <cassert>
+#include <cstddef>
+#include <cstdio>
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <string_view>
+#include <thread>
+#include <vector>
+
+#include "debug_logging.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest-spi.h"
+#include "gtest/gtest.h"
+
+using ::testing::ElementsAre;
+using ::testing::TempDir;
+
+namespace performancelayers {
+namespace {
+constexpr unsigned kLogsPerThread = 1337;
+constexpr unsigned kThreadCount = 100;
+
+// An abstraction over the output. It writes the incoming string to the output.
+class LogOutput {
+ public:
+  virtual ~LogOutput() = default;
+
+  virtual void Flush() = 0;
+
+  virtual void LogLine(const std::string_view line) = 0;
+};
+
+// Implements LogOutput for a file. If the given filename is `nullptr`, it
+// writes to the standard output. After each write, the logs are persisted to
+// the file by calling `fflush`. Since only one line is written to the output by
+// each `LogLine()` call, it doens't need to aquire a mutex.
+class FileOutput : public LogOutput {
+ public:
+  FileOutput(const char *filename) {
+    if (!filename) {
+      out_ = stderr;
+      return;
+    }
+
+    // Since multiple layers open the same file and write to it at the same
+    // time, it's opened in the append mode.
+    out_ = fopen(filename, "a");
+    if (!out_) {
+      SPL_LOG(ERROR) << "Failed to open " << filename
+                     << ". Using stderr as the alternative output.";
+      out_ = stderr;
+    }
+  }
+
+  ~FileOutput() {
+    if (out_ && out_ != stderr) {
+      fclose(out_);
+    }
+  }
+
+  void Flush() override {
+    assert(out_);
+    fflush(out_);
+  }
+
+  void LogLine(const std::string_view line) override {
+    assert(out_);
+    assert(line.find('\n') == std::string_view::npos &&
+           "Expected single line.");
+    fprintf(out_, "%s\n", line.data());
+    Flush();
+  }
+
+ private:
+  FILE *out_ = nullptr;
+};
+
+// This class is used for testing. It writes the data to a string
+// instead of a file. The data can be read using the `GetLog()` method.
+class StringOutput : public LogOutput {
+ public:
+  StringOutput(){};
+
+  void Flush() override {}
+
+  void LogLine(const std::string_view line) override {
+    assert(line.find('\n') == std::string_view::npos &&
+           "Expected single line.");
+    out_.push_back(line);
+  }
+
+  const std::vector<std::string_view> &GetLog() { return out_; }
+
+ private:
+  std::vector<std::string_view> out_;
+};
+
+// Used by the tests prior to logging to make sure the underlying file is empty.
+void CreateEmptyFile(std::string &filename) { std::ofstream(filename.c_str()); }
+
+TEST(LogOutput, SingleThreadedLog) {
+  std::string file_path = TempDir() + "/single_thread.log";
+  CreateEmptyFile(file_path);
+
+  const std::string line = "name:event_name,timestamp:1234";
+  FileOutput file_out(file_path.c_str());
```

Comment:
I don't think you need `c_str`? This would be slow because `string_view` would have to call `strlen` on the c-string instead of using the known length. Instead, either convert it directly to `string_view` or make `line` itself a `string_view`

Reply:
disregard, this creates a new file from the path